x-pack/filebeat/module/cisco/asa/ingest/pipeline.json

```diff
@@ -0,0 +1,266 @@
+        {"set"    : {"if": "ctx.cisco.asa.message_id == '338301'", "field": "client.port", "value": "{{destination.port}}"}},
+        {"set"    : {"if": "ctx.cisco.asa.message_id == '338301'", "field": "server.ip", "value": "{{source.ip}}"}},
+        {"set"    : {"if": "ctx.cisco.asa.message_id == '338301'", "field": "server.port", "value": "{{source.port}}"}},
+
+        {<- /* handle 302xxx messages (Flow expiration a.k.a "Teardown") */ ->}
+        {
+            "grok": {
+                "field": "log.original",
+                "if": "[\"302014\", \"302016\", \"302018\", \"302021\", \"302036\", \"302304\", \"302306\"].contains(ctx.cisco.asa.message_id)",
+                "patterns": [
+                    "Teardown %{NOTSPACE:network.transport} (:?state-bypass )?connection %{NOTSPACE:cisco.asa.connection_id} (:?for|from) %{NOTCOLON:cisco.asa.source_interface}:%{IP:source.ip}\/%{NUMBER:source.port:int} (:?%{NOTSPACE:cisco.asa.source_username} )?to %{NOTCOLON:cisco.asa.destination_interface}:%{IP:destination.ip}\/%{NUMBER:destination.port:int} (:?%{NOTSPACE:cisco.asa.destination_username} )?(:?duration %{TIME:_temp_.duration_hms} bytes %{NUMBER:network.bytes:int})%{GREEDYDATA}",
+                    "Teardown %{NOTSPACE:network.transport} connection for faddr (:?%{NOTCOLON:cisco.asa.source_interface}:)?%{IP:source.ip}/%{NUMBER:source.port:int} (:?%{NOTSPACE:cisco.asa.source_username} )?gaddr (:?%{NOTCOLON}:)?%{IP}/%{NUMBER} laddr (:?%{NOTCOLON:cisco.asa.destination_interface}:)?%{IP:destination.ip}/%{NUMBER:destination.port:int}(:? %{NOTSPACE:cisco.asa.destination_username})?%{GREEDYDATA}"
+                ],
+                "pattern_definitions": {
+                    "NOTCOLON": "[^:]*"
+                }
+            }
+        },
+        {
+            "set": {
+                "if": "[\"302014\", \"302016\", \"302018\", \"302021\", \"302036\", \"302304\", \"302306\"].contains(ctx.cisco.asa.message_id)",
+                "field": "event.action",
+                "value": "flow-expiration"
+            }
+        },
+
+        {<- /* process the flow duration "hh:mm:ss" present in some messages */ ->}
+
+        {
+            "script": {
+                "lang": "painless",
+                "if": "ctx?._temp_?.duration_hms != null",
+                "source": "{<- /* This script sets event.{start,end,duration}: */ ->}
+                           {<- /* - event.end is set from @timestamp           */ ->}
+                           {<- /* - event.duration is the result of parsing    */ ->}
+                           {<- /*   a string in the form "hh:mm:ss"            */ ->}
+                           {<- /* - event.start is event.end - event.duration. */ ->}
+                           {<- "long parse_hms(String s) {                        ">}
+                           {<- "    long cur = 0;                                 ">}
```

Comment:
Interesting use of `text/template`. Did you try making it accept YAML? Optimistically I wonder if changing this one line would make it work.

https://github.com/elastic/beats/blob/7542db1cb199a55ac9766ed6021a8dbaa81bc3bc/filebeat/fileset/fileset.go#L430